File: src-tauri/src/commands/serial_script.rs
// Copyright 2024 Serial CLI Contributors
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

use crate::state::app_state::AppState;
use serial_cli::lua::ui_actions::UiAction;
use serial_cli::serial_core::serial_script::SerialScriptEngine;
use tauri::State;

/// Attach a Lua script engine to an open serial port.
/// The script's lifecycle hooks (on_open, on_send, on_recv, on_timer, on_close)
/// will be called automatically during port operations.
#[tauri::command]
pub async fn attach_script(
    port_id: String,
    script_source: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let engine =
        SerialScriptEngine::new(&script_source).map_err(|e| format!("Invalid script: {}", e))?;

    let manager = state.port_manager.lock().await;
    manager
        .attach_script(&port_id, engine)
        .await
        .map_err(|e| e.to_string())
}

/// Detach the Lua script engine from an open serial port.
/// Calls `on_close` and stops the timer before detaching.
#[tauri::command]
pub async fn detach_script(port_id: String, state: State<'_, AppState>) -> Result<(), String> {
    let manager = state.port_manager.lock().await;
    manager
        .detach_script(&port_id)
        .await
        .map_err(|e| e.to_string())
}

/// Check whether a serial port has a script engine attached.
#[tauri::command]
pub async fn has_script(port_id: String, state: State<'_, AppState>) -> Result<bool, String> {
    let manager = state.port_manager.lock().await;
    manager
        .has_script(&port_id)
        .await
        .map_err(|e| e.to_string())
}

/// Get the script status for a serial port.
/// Returns `{ has_script, timer_interval_ms }`.
#[tauri::command]
pub async fn get_script_status(
    port_id: String,
    state: State<'_, AppState>,
) -> Result<ScriptStatus, String> {
    let manager = state.port_manager.lock().await;
    let (has, timer_ms) = manager
        .get_script_status(&port_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(ScriptStatus {
        has_script: has,
        timer_interval_ms: timer_ms,
    })
}

/// List all UI actions (`action_*` functions) exposed by the script attached to a port.
#[tauri::command]
pub async fn list_script_actions(
    port_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<UiAction>, String> {
    let manager = state.port_manager.lock().await;
    manager
        .list_script_actions(&port_id)
        .await
        .map_err(|e| e.to_string())
}

/// Execute a UI action function on the port's attached script engine.
#[tauri::command]
pub async fn call_script_function(
    port_id: String,
    function_name: String,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let manager = state.port_manager.lock().await;
    manager
        .call_script_action(&port_id, &function_name)
        .await
        .map_err(|e| e.to_string())
}

/// Script status response
#[derive(serde::Serialize)]
pub struct ScriptStatus {
    pub has_script: bool,
    pub timer_interval_ms: u64,
}

File: frontend/src/stores/serialScript.ts
import { create } from "zustand";
import { tauriApi } from "@/lib/tauri-api";
import type { ScriptStatus, UiAction } from "@/types";

interface SerialScriptStore {
  attachedScript: string | null;
  scriptStatus: ScriptStatus | null;
  actions: UiAction[];
  loading: boolean;

  attachScript: (portId: string, scriptSource: string) => Promise<void>;
  detachScript: (portId: string) => Promise<void>;
  refreshStatus: (portId: string) => Promise<void>;
  loadActions: (portId: string) => Promise<void>;
  callAction: (portId: string, functionName: string, args?: string) => Promise<string>;
}

interface StandaloneScriptStore {
  actions: UiAction[];
  currentScriptSource: string;
  loading: boolean;
  output: string | null;

  loadActions: (scriptSource: string) => Promise<void>;
  callAction: (functionName: string, args?: string) => Promise<string>;
}

export const useSerialScriptStore = create<SerialScriptStore>()((set) => ({
  attachedScript: null,
  scriptStatus: null,
  actions: [],
  loading: false,

  attachScript: async (portId, scriptSource) => {
    set({ loading: true });
    try {
      await tauriApi.attachScript(portId, scriptSource);
      set({ attachedScript: scriptSource, loading: false });
    } catch {
      set({ loading: false });
    }
  },

  detachScript: async (portId) => {
    set({ loading: true });
    try {
      await tauriApi.detachScript(portId);
      set({
        attachedScript: null,
        scriptStatus: null,
        actions: [],
        loading: false,
      });
    } catch {
      set({ loading: false });
    }
  },

  refreshStatus: async (portId) => {
    try {
      const status = await tauriApi.getScriptStatus(portId);
      set({ scriptStatus: status });
    } catch {
      // ignore
    }
  },

  loadActions: async (portId) => {
    try {
      const actions = await tauriApi.listScriptActions(portId);
      set({ actions });
    } catch {
      set({ actions: [] });
    }
  },

  callAction: async (portId, functionName, args) => {
    return await tauriApi.callScriptFunction(portId, functionName, args);
  },
}));

export const useStandaloneScriptStore = create<StandaloneScriptStore>()(
  (set, get) => ({
    actions: [],
    currentScriptSource: "",
    loading: false,
    output: null,

    loadActions: async (scriptSource) => {
      set({ loading: true, currentScriptSource: scriptSource });
      try {
        const actions =
          await tauriApi.listStandaloneScriptActions(scriptSource);
        set({ actions, loading: false });
      } catch {
        set({ actions: [], loading: false });
      }
    },

    callAction: async (functionName, args) => {
      const { currentScriptSource } = get();
      const result = await tauriApi.callStandaloneScriptFunction(
        currentScriptSource,
        functionName,
        args,
      );
      set({ output: result });
      return result;
    },
  }),
);
